mysql-test/suite/binlog/t/binlog_recover_async_rollback_trx.test

```diff
@@ -0,0 +1,82 @@
+--source include/have_debug.inc
+--source include/have_innodb.inc
+--source include/have_binlog_format_row.inc
+
+CREATE TABLE t1 (id int primary key, data int) ENGINE = InnoDB;
+INSERT INTO t1 VALUES (0, 1);
+
+--echo #
+--echo # 1. Check DML in prepared state can rollback correctly.
+--echo #
+
+--source include/expect_crash.inc
+SET debug = "d,crash_before_writing_xid";
+--error 2013
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+
+--let $restart_parameters= --debug=d,print_log_after_rollback_on_recovery
+--source include/start_mysqld.inc
+
+--echo # Check whether the prepared transaction of UPDATE has been rollback.
+--echo # Expect (0,1)
+SELECT * FROM t1;
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+# Expect (0,2)
+SELECT * FROM t1;
+
+# Check whether the DML trx has been rolled back asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server reported: recovery_rollback_thread
+--let $assert_select= recovery_rollback_thread
+--let $assert_count= 1
+--source include/assert_grep.inc
+
+--echo #
+--echo # 2. Check DML can rollback automatically after second crash.
+--echo #
+
+--source include/expect_crash.inc
+SET debug = "d,crash_before_writing_xid";
+--error 2013
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+
+# Restart the server.
+--let $restart_parameters= --debug=d,print_log_after_rollback_on_recovery,block_rollback_thread
+--source include/start_mysqld.inc
+
+SELECT COUNT(*) FROM information_schema.innodb_trx WHERE trx_state = "RUNNING";
+# Expect(0,2)
+SELECT * FROM t1;
+SET TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
+# Expect(0,3)
+SELECT * FROM t1;
+
+# Crash again.
+--source include/expect_crash.inc
+SET debug = "d,crash_dispatch_command_before";
+--error 2013
+SELECT 1;
+
+# Restart the server.
+--let $restart_parameters = --debug=d,print_log_after_rollback_on_recovery
+--source include/start_mysqld.inc
+
+# Check whether the prepared transaction has been rollback.
+# Expect(0,2)
+SELECT * FROM t1;
+UPDATE t1 SET data = data + 1 WHERE id = 0;
+# Expect(0,3)
+SELECT * FROM t1;
+
+# Check whether the DML trx have been rollback asynchronously.
+# For the first crash recovery, prepare transaction become
+# active, no log print because rollback not finish.
+# For the second crash recovery, rollback active trx asynchronously.
+--let $assert_file= $MYSQLTEST_VARDIR/log/mysqld.1.err
+--let $assert_text= Server reported: recovery_rollback_thread
+--let $assert_select= recovery_rollback_thread
+--let $assert_count= 2
+--source include/assert_grep.inc
```

Comment:
I would suggest to use `include/search_pattern_in_file.inc` instead.

Reply:
Any reason to use search_pattern_in_file.inc instead of assert_grep.inc? I used assert_grep.inc, because I want the test stops immediately if any error happens.

Reply:
Because there are multiple test steps that seem independent of each other, the rest of the test would be unnecessarily ruined because of the first failure. Replication tests are known to fail intermittently.

Reply:
got it. I used a separate log file to avoid what you said.